src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -314,14 +362,36 @@ private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byt
 
             if (headerSpan.Length < 5)
             {
-                // Less than min possible headerSpan is 5 bytes a:b\r\n
-                RejectRequestHeader(headerSpan);
+                if (!_enableLineFeedTerminator || headerSpan.Length < 4)
+                {
+                    // Less than min possible headerSpan is 4 bytes a:b\n or 5 bytes a:b\r\n
+                    RejectRequestHeader(headerSpan);
+                }
             }
 
             if (headerSpan[^2] != ByteCR)
             {
-                // Sequence needs to be CRLF not LF first.
-                RejectRequestHeader(headerSpan[..^1]);
+                if (!_enableLineFeedTerminator)
+                {
+                    // Sequence needs to be CRLF.
+                    RejectRequestHeader(headerSpan[..^1]);
+                }
+                else if (headerSpan[^2] != ByteLF)
+                {
+                    RejectRequestHeader(headerSpan[..^1]);
+                }
+                else
+                {
+                    // Line ends with LF and quirk mode is enabled, try to parse the header
+                    if (!TryTakeSingleHeader(handler, headerSpan[..^1]))
+                    {
+                        // Not parsable as a valid name:value header pair.
+                        RejectRequestHeader(headerSpan);
+                    }
+                    
+                    // Omit last byte since the header is a valid LF terminated line
+                    return headerSpan.Length - 1;
```

Comment:
Why are we omitting the last byte here? Don't we want to consume the LF? We don't return `headerSpan.Length - 2` in the CRLF case.